### PR TITLE
feat: remember me option

### PR DESCRIPTION
**Pull request by aurelien-pacaud**
## :id: Reference related issue. 
https://gravitee.atlassian.net/browse/AM-744

## :pencil2: A description of the changes proposed in the pull request

If the user checks "remember me" during sign in a cookie is created to keep the user authenticated. Even if the session expired, if a valid remember cookie exists the user will be authenticated.

In production it's important to define a remember me duration greater than the session timeout.

## :memo: Test scenarios 

To be able to test and not wait 30 min it's important not persist session after closing the browser.

**Default sign in with configuration enabled at domain level**

- Create an application
- Register a user
- Enable the configuration to change the remember me duration at domain level (Settings -> User accounts)
- Choose a duration
- Disable persistent session (Application -> Settings -> Session management) 
- Sign in with the user with "remember me" checked
- Close your browser, the remember me cookie will be keep, the cookie session is removed
- Wait to let the remember me cookie expired
- Call again `oauth/authorize`
- The login form is displayed

**Default sign in with configuration enabled at application level**

- Create an application
- Register a user
- Enable the configuration to change the session duration at domain level (Your application -> Settings -> User accounts)
- Choose a duration
- Disable persistent session (Your application -> Settings -> Session management) 
- Sign in with the user with "remember me" checked
- Close your browser, the remember me cookie will be keep, the cookie session is removed
- Wait to let the remember me cookie expired
- Call again `oauth/authorize`
- The login form is displayed

**Webauth sign**

- Create an application
- Register a user
- Configure webauth at domain level
- Enable the configuration to change the session duration at domain or application level
- Choose a duration
- Click on 'Sign in with fingerprint, device or security key'
- Sign in with the user with "remember me" checked
- Close your browser, the remember me cookie is kept, the cookie session is removed
- Wait to let the remember me cookie expired
- Call again `oauth/authorize`
- The login form is displayed

**Identifier first**

- Create an application
- Register a user
- Configure 'identifier first'
- Enable the configuration to change the session duration at domain or application level
- Choose a duration
- Sign in with the user with "remember me" checked
- Close your browser, the remember me cookie will be keep, the cookie session is removed
- Wait to let the remember me cookie expired
- Call again `oauth/authorize`
- The login form is displayed

**Identifier first with social login**

- Create an application
- Register a user
- Configure 'identifier first'
- Enable the configuration to change the session duration at domain or application level
- Choose a duration
- Create a social IDP with Google for example
- Enable this IDP in your application
- Add a "Selection rule" to this IDP (username matching for example)
- Sign in with the user with "remember me" checked
- You will be automatically redirected to the IDP to authenticate the user
- Close your browser, the remember me cookie will be keep, the cookie session is removed
- Wait to let the remember me cookie expired
- Call again `oauth/authorize`
- The login form is displayed

## :computer: Add screenshots for UI

Screenshot 2023-08-29 at 14 24 38
Screenshot 2023-08-29 at 14 24 28
Screenshot 2023-08-29 at 14 24 12

### CODE REVIEW

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/resources/liquibase/changelogs/v4_1_0/4.1.0-update-users-table-remember-me.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+databaseChangeLog:
```
Comment (tcompiegne) [inline gravitee-am-repository/gravitee-am-repository-jdbc/src/main/resources/liquibase/changelogs/v4_1_0/4.1.0-update-users-table-remember-me.yml:1]: This file is not used anymore.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java [hunk 2]
```diff
@@ -284,6 +285,9 @@ public class RootProvider extends AbstractService<ProtocolProvider> implements P
     @Autowired
     private EmailService emailService;
 
+    @Autowired
+    private io.gravitee.am.service.UserService userService;
```
Comment (tcompiegne) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:289]: Not sure this service is used.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java [hunk 3]
```diff
@@ -209,7 +210,7 @@ public class RootProvider extends AbstractService<ProtocolProvider> implements P
 
     @Autowired
     @Qualifier("managementUserService")
-    private UserService userService;
+    private UserService managementuserService;
```
Comment (tcompiegne) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:213]: I think we can revert back to userService name.

**Review by tcompiegne — CHANGES_REQUESTED** [review]

File: gravitee-am-model/src/main/java/io/gravitee/am/model/account/AccountSettings.java [hunk 4]
```diff
@@ -107,6 +107,16 @@ public class AccountSettings {
...
+    /**
+     * Remember me option, to avoid to sign in again during a predefined amount of time
+     */
+    private boolean rememberMe;
+
+    /**
+     * Amount of time while the cookie session is stored
+     */
+    private Integer rememberMeDuration;
+
```
[the remark this answers is not in the document]
Comment (aurelien-pacaud) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/account/AccountSettings.java:119]: I just saw that but in the specification/figma it has to be configurable from "user accounts"

**Review by tcompiegne — APPROVED** [review]

**Pull request merged**